linked_list/linked_list.py

```diff
@@ -1,97 +1,197 @@
+        while current_node:
+            if current_node.value == target_node.value:
+                return True
+            current_node = current_node.next
+        return False
 
     # method that returns the length of the singly linked list
     # Time Complexity: ?
     # Space Complexity: ?
     def length(self):
-        pass
+        if self.head is None:
+            return 0
+        current_node = self.head
+        length_of_list = 0
+
+        # traverse list til reach end, increment length while traversing.
+        while current_node:
+            length_of_list += 1
+            current_node = current_node.next
+
+        return length_of_list
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: linear
+    # Space Complexity: constant
     def get_at_index(self, index):
-        pass
+        if index < 0 or self.length() < index:
+            return None
+        current_index = 0
+        current_node = self.head
+
+        # traverse list till reach index or end of list
+        while current_node:
+            # can add a condition to check the current node is less than index if above does not work
+            if current_index == index:
+                break
+            current_node = current_node.next
+            current_index += 1
+
+        return current_node.value
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: linear
+    # Space Complexity: constant
     def get_last(self):
-        pass
+        if self.head is None:
+            return None
+        current_node = self.head
+        tail_node = current_node
+        prev_node = None
+
+        # traverse list till reach end.
+        while current_node:
+            prev_node = current_node
+            current_node = current_node.next
+        tail_node = prev_node
+        return tail_node.value
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: linear
+    # Space Complexity: constant
     def add_last(self, value):
-        pass
+        if self.head is None:
+            self.head = value
+            return
+
+        prev_node = None
+        current_node = self.head
+        new_tail_node = Node(value)
+
+        # traverse list till reach end.
+        while current_node:
+            prev_node = current_node
+            current_node = current_node.next
+        # re-assign nodes so last node we found becomes second to last pointing to new tail and new tail points to none
+        new_tail_node = prev_node
+        current_node.next = new_tail_node
+        new_tail_node.next = None

```

Comment:
Here, after creating the `new_tail_node` on line 115, you are overwriting it with `prev_node` which, once the while loop is finished executing, should be the current tail node. 

Once the while loop is finished executing `current_node` is going to be the node after the the current tail, aka `None`. 

Instead, try pointing the current tail (`prev_node`) to `new_tail_node`